File: day16_part2_2020.rb

def parse_input(file)
  sections = File.read(file).split("\n\n")
  rules = sections[0].split("\n").map do |line|
    field, ranges = line.split(": ")
    {
      field: field,
      ranges: ranges.split(" or ").map { |range| range.split("-").map(&:to_i) }
    }
  end
  my_ticket = sections[1].split("\n")[1].split(",").map(&:to_i)
  nearby_tickets = sections[2].split("\n")[1..].map { |line| line.split(",").map(&:to_i) }
  [rules, my_ticket, nearby_tickets]
end

def valid_value?(value, rules)
  rules.any? { |rule| rule[:ranges].any? { |(min, max)| value.between?(min, max) } }
end

def part_one(rules, nearby_tickets)
  nearby_tickets.flatten.select { |value| !valid_value?(value, rules) }.sum
end

def valid_ticket?(ticket, rules)
  ticket.all? { |value| valid_value?(value, rules) }
end

def determine_fields(rules, valid_tickets)
  possible_fields = Array.new(rules.size) { rules.map { |rule| rule[:field] } }
  valid_tickets.each do |ticket|
    ticket.each_with_index do |value, index|
      possible_fields[index].select! do |field|
        rule = rules.find { |rule| rule[:field] == field }
        rule[:ranges].any? { |(min, max)| value.between?(min, max) }
      end
    end
  end
  until possible_fields.all? { |fields| fields.size == 1 }
    confirmed_fields = possible_fields.select { |fields| fields.size == 1 }.flatten
    possible_fields.each { |fields| fields.reject! { |field| confirmed_fields.include?(field) } if fields.size > 1 }
  end
  possible_fields.flatten
end

def part_two(rules, my_ticket, valid_tickets)
  fields_order = determine_fields(rules, valid_tickets)
  departure_values = fields_order.each_with_index.map { |field, index| my_ticket[index] if field.start_with?("departure") }.compact
  departure_values.reduce(:*)
end

rules, my_ticket, nearby_tickets = parse_input("input.txt")
valid_tickets = nearby_tickets.select { |ticket| valid_ticket?(ticket, rules) }

puts part_one(rules, nearby_tickets) # Output for Part One
puts part_two(rules, my_ticket, valid_tickets) # Output for Part Two

File: training_data/reflection-tuning/day16_part2_2020.py
import re
from math import prod

def parse_input(filename):
    with open(filename, 'r') as f:
        rules_raw, your_ticket_raw, nearby_tickets_raw = f.read().split('\n\n')

    rules = {}
    for line in rules_raw.split('\n'):
        field, ranges = line.split(': ')
        rules[field] = [tuple(map(int, r.split('-'))) for r in ranges.split(' or ')]

    your_ticket = list(map(int, your_ticket_raw.split('\n')[1].split(',')))
    nearby_tickets = [list(map(int, line.split(','))) for line in nearby_tickets_raw.split('\n')[1:]]

    return rules, your_ticket, nearby_tickets

def is_valid_for_any_field(value, rules):
    return any(any(low <= value <= high for low, high in ranges) for ranges in rules.values())

def solve(filename):
    rules, your_ticket, nearby_tickets = parse_input(filename)

    # Part 1
    error_rate = sum(value for ticket in nearby_tickets for value in ticket if not is_valid_for_any_field(value, rules))
    print(f"Part 1: {error_rate}")

    # Part 2
    valid_tickets = [ticket for ticket in nearby_tickets if all(is_valid_for_any_field(value, rules) for value in ticket)]

    possible_fields = [set(rules.keys()) for _ in range(len(your_ticket))]
    for ticket in valid_tickets:
        for i, value in enumerate(ticket):
            possible_fields[i] = {field for field in possible_fields[i] if any(low <= value <= high for low, high in rules[field])}

    field_positions = {}
    while len(field_positions) < len(rules):
        for i, fields in enumerate(possible_fields):
            if len(fields) == 1:
                field = fields.pop()
                field_positions[field] = i
                for other_fields in possible_fields:
                    other_fields.discard(field)

    departure_values = [your_ticket[pos] for field, pos in field_positions.items() if field.startswith('departure')]
    result = prod(departure_values)
    print(f"Part 2: {result}")

solve('input.txt')
